Remove helpers from Rails APIs

djsegal opened the pull request:
This is two changes:

+ extending the https://github.com/plataformatec/devise/pull/3732 logic to devise_controller ( i.e. `respond_to?(:helper_method)` )
+ emptying `@@helpers` for Rails Api's

----------

**see error below**
the reason i need to empty `@@helpers` is because I'm using `DeviseTokenAuth` and just getting rid of `Devise::Controllers::Helpers` inside the codebase would still leave that one helper left (which sparks errors)

```
[1] pry(Devise)> @@helpers
=> #<Set: {Devise::Controllers::Helpers, DeviseTokenAuth::Controllers::Helpers}>
```

djsegal commented:
this builds off https://github.com/plataformatec/devise/pull/3714

twalpole commented:
Can you add testcases?

twalpole commented on lib/devise.rb:
```diff
@@ -325,6 +325,9 @@ def self.add_mapping(resource, options)
     @@mappings[mapping.name] = mapping
     @@default_scope ||= mapping.name
+    if defined? Rails and Rails.try(:application).try(:config).try(:api_only)
+      @@helpers = []
```
I don't fully understand why this is needed - but by default @@helpers is a Set - so, if it is necessary to clear it, it's probably better to set to Set.new rather than empty array,  it may also be better to not call define_helpers if API rather than clearing @@helpers (If that fixes the issue this is needed for)

djsegal commented:
@twalpole: i just updated the commit. 
- I dropped the hacky `@@helpers = []` because I realized it was something I could fix in a PR to `devise_token_auth`
- I added a test that checks for helper methods. I don't know how to go about making an integration test, though

djsegal commented:
@twalpole sorry about this. closing the issue. resolved by https://github.com/lynndylanhurley/devise_token_auth/pull/469

djsegal commented:
@twalpole sorry about this. reopening this issue. partially resolved by https://github.com/lynndylanhurley/devise_token_auth/pull/469, but still needs the check on `helper` and `helper_method` from this PR

twalpole commented:
@djsegal  This branch was merged into the official devise master -- you probably want to issue this PR against that.
